[go1.16] Update to go1.16

justaugustus opened the pull request:
**What type of PR is this?**

/kind feature
/area dependency

**What this PR does / why we need it**:

- Update to go1.16

Tracking issue: https://github.com/kubernetes/release/issues/1834
Slack thread: https://kubernetes.slack.com/archives/C2C40FMNF/p1610074100305000
cc: @kubernetes/release-engineering 

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
- Kubernetes is now built with Golang 1.16
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

justaugustus committed:
[go1.16] Update to go1.16

Signed-off-by: Stephen Augustus <foo@auggie.dev>

justaugustus committed:
[go1.16] staging/publishing: Set default go version to go1.16

Signed-off-by: Stephen Augustus <foo@auggie.dev>

hasheddan reviewed (approved):
/lgtm

justaugustus commented:
/assign @dims 
FYI @nikhita

justaugustus commented:
(now why did I think this was going to be simple...)
/retest

dims commented:
@justaugustus is the intention to merge the RC1 or wait for the final release ( which is probably in a few days 🤞🏾 )

justaugustus commented:
> @justaugustus is the intention to merge the RC1 or wait for the final release ( which is probably in a few days 🤞🏾 )

I was hoping to get the RC1 in, but we'll see what the tests have to say about that.
Depending on how timing lines up, this may just end up being the go1.16.0 PR.

justaugustus committed:
[go1.16] Use go-runner:v2.3.1-go1.16-buster.0 image

Signed-off-by: Stephen Augustus <foo@auggie.dev>

justaugustus commented on staging/publishing/rules.yaml:
```diff
@@ -4,7 +4,7 @@ recursive-delete-patterns:
 - "*/BUILD.bazel"
 - Gopkg.toml
-default-go-version: 1.15.7
+default-go-version: 1.16rc1
```
@dims @nikhita -- Have you seen [this](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98572/pull-publishing-bot-validate/1355249121295863808) crop up w/ pre-release versions before?

```console
 F0129 20:22:31.932516     163 main.go:35] Invalid rules file "/home/prow/go/src/k8s.io/kubernetes/staging/publishing/rules.yaml": validation errors:
- specified go version 1.16rc1 must be a valid go version: Invalid character(s) found in minor number "16rc1" 
```

dims replied:
@justaugustus we validate using blang/semver, that's what is failing. See snippet https://play.golang.org/p/gNIlfcIoibb

nikhita replied:
Created https://github.com/kubernetes/publishing-bot/pull/245

nikhita replied:
`pull-publishing-bot-validate` is now green :tada:

nikhita commented:
/test pull-publishing-bot-validate

nikhita commented:
Hmm, looks like `pull-publishing-bot-validate` is not fetching the new publishing-bot code.

nikhita commented:
/test pull-publishing-bot-validate

dashpole commented:
/remove-sig instrumentation

justaugustus committed:
[go1.16] go.mod: update to go1.16

Signed-off-by: Stephen Augustus <foo@auggie.dev>

justaugustus commented:
/retest

fedebongio commented:
/remove-sig api-machinery

wzshiming commented:
https://golang.org/dl/
go 1.6 has been released whether to use it

dashpole commented:
/remove-sig instrumentation

fedebongio commented:
/remove-sig api-machinery

justaugustus committed:
generated: Run hack/lint-dependencies.sh and hack/update-vendor.sh

Signed-off-by: Stephen Augustus <foo@auggie.dev>

liggitt committed:
[go1.16] bump golang.org/x/... dependencies

hack/pin-dependency.sh golang.org/x/crypto latest
hack/pin-dependency.sh golang.org/x/net latest
hack/pin-dependency.sh golang.org/x/exp latest
hack/pin-dependency.sh golang.org/x/sys latest
hack/pin-dependency.sh golang.org/x/time latest
hack/pin-dependency.sh golang.org/x/tools latest

hack/lint-dependencies.sh
hack/pin-dependency.sh dmitri.shuralyov.com/gpu/mtl v0.0.0-20201218220906-28db891af037
hack/pin-dependency.sh golang.org/x/mobile v0.0.0-20201217150744-e6ae53a27f4f
hack/pin-dependency.sh golang.org/x/mod v0.3.1-0.20200828183125-ce943fd02449

hack/lint-dependencies.sh

hack/update-internal-modules.sh
hack/update-vendor.sh

Co-authored-by: Stephen Augustus <foo@auggie.dev>

fedebongio commented:
/remove-sig api-machinery

justaugustus commented:
Rebased to remove the bazel-related commits.
Shoutout to @BenTheElder! 💕 
ref: https://github.com/kubernetes/kubernetes/pull/99561

justaugustus commented:
The [dependencies test is failing](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98572/pull-kubernetes-dependencies/1366533122371358720) on verify-vendor.sh, which I believe is a result of the go.mod/go.sum diffs between go1.15 and go1.16.

`gcr.io/k8s-testimages/kubekins-e2e:v20210226-c001921-master` is built on go1.15, while this PR should be tested w/ go1.16.

justaugustus commented:
/test pull-kubernetes-node-e2e

justaugustus commented:
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

ref: https://github.com/kubernetes/test-infra/pull/21107

justaugustus commented:
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

ref: https://github.com/kubernetes/test-infra/pull/21108

liggitt commented:
all the bits in this PR lgtm

do we have a canary integration job that uses go1.16?

justaugustus committed:
Fixup golang.org/x/term staticcheck warnings

Errors from staticcheck:
cmd/preferredimports/preferredimports.go:38:2:
  package golang.org/x/crypto/ssh/terminal is deprecated:
  this package moved to golang.org/x/term.  (SA1019)
vendor/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:36:2:
  package golang.org/x/crypto/ssh/terminal is deprecated:
  this package moved to golang.org/x/term.  (SA1019)
vendor/k8s.io/client-go/tools/clientcmd/auth_loaders.go:26:2:
  package golang.org/x/crypto/ssh/terminal is deprecated:
  this package moved to golang.org/x/term.  (SA1019)

Please review the above warnings. You can test via:
  hack/verify-staticcheck.sh <failing package>
If the above warnings do not make sense, you can exempt the line or
file. See:
  https://staticcheck.io/docs/#ignoring-problems

generated:
- hack/update-internal-modules.sh
- hack/lint-dependencies.sh
- hack/update-vendor.sh

Signed-off-by: Stephen Augustus <foo@auggie.dev>

justaugustus commented:
> do we have a canary integration job that uses go1.16?

@liggitt -- Opened a PR to add this: https://github.com/kubernetes/test-infra/pull/21110.

justaugustus commented:
/test pull-kubernetes-node-e2e
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

justaugustus commented:
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

justaugustus commented:
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

justaugustus commented:
/test pull-kubernetes-dependencies-go-canary

justaugustus commented:
I needed to fix up some staticcheck warnings in https://github.com/kubernetes/kubernetes/pull/98572/commits/d9435800b65d6787beaf061546599d757b8e87c9, but the following go1.16 presubmits are now passing:
- `pull-kubernetes-dependencies-go-canary`: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98572/pull-kubernetes-dependencies-go-canary/1366634510489227264
- `pull-kubernetes-verify-go-canary`: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98572/pull-kubernetes-verify-go-canary/1366631602737647616

I have some confidence that the remaining tests currently running (pull-kubernetes-e2e-azure-disk-windows, pull-kubernetes-e2e-azure-file-windows-containerd, pull-kubernetes-e2e-gce-100-performance, pull-kubernetes-e2e-gce-ubuntu-containerd) will pass.

Once they do, https://github.com/kubernetes/test-infra/pull/21109 will need to be approved and merged, which should get the `pull-kubernetes-dependencies` and `pull-kubernetes-verify` jobs to go green.

justaugustus commented:
Or not 🙃 
/test pull-kubernetes-e2e-azure-disk-windows

justaugustus commented:
https://github.com/kubernetes/test-infra/pull/21109 has merged:
/test pull-kubernetes-dependencies
/test pull-kubernetes-verify

justaugustus commented:
/assign @liggitt

justaugustus commented:
All green now! ✨ 🌈

saschagrunert reviewed (approved):
/lgtm

cpanato reviewed (approved):
/lgtm

justaugustus commented:
/test pull-kubernetes-integration-go-canary

justaugustus commented:
`pull-kubernetes-integration-go-canary` passed as well: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98572/pull-kubernetes-integration-go-canary/1366759123974098944

liggitt commented:
/approve

liggitt commented:
/triage accepted
/priority important-soon

justaugustus commented:
![Screenshot from 2021-03-02 10-56-22](https://user-images.githubusercontent.com/567897/109675726-0c58b780-7b46-11eb-8097-d5703b32a066.png)

/label wake-up-bots
/honk
